tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,279 @@
+
+#enable 8 virtual functions for the specified physical functions of the SR-IOV network device 
+sub enable_vf(){
+  my @pfs = @_;
+  
+  # get the network device drivers
+  my (@drivers, $pf, $driver);
+  foreach $pf (@pfs) {
+    $driver = script_output "lspci -v -s $pf | sed -n '/Kernel modules/p' | sed 's/.*Kernel modules: *//'";
+    my $is_known_driver = 0;
+    foreach (@drivers) {
+      next if ($_ ne "$driver");
+      $is_known_driver = 1;
+    }
+    push @drivers, $driver if ! $is_known_driver;
+  }
+  print "julies: \@drivers=@drivers\n";
+  foreach (@drivers) {
+    assert_script_run("modprobe $driver", 60);
+  }
+ 
+  # set max_vfs and reload driver
+  #it seems the repeated enabling vf consused the system, or the some vf is not detached from a shutdown vm?
+  #so skip enabling in local test for debugging
+  #record_info("Skip enabling VF");
+  foreach $driver (@drivers) {
+    assert_script_run("rmmod $driver", 30);   #4s last time
+    assert_script_run("modprobe --first-time $driver max_vfs=8", 60);   #2s last time. but once fail with 30
+  }
+
+  #bring up the SR-IOV device
+  foreach $pf (@pfs) {
+    my $nic = script_output "ls -l /sys/class/net |grep $pf | awk '{print \$9}'";
+    assert_script_run "echo \"BOOTPROTO='dhcp'\nSTARTMODE='manual'\" > /etc/sysconfig/network/ifcfg-$nic";
+    assert_script_run("ifup $nic", 60);   #15s last time
+  }
+  
+  my $vfs = script_output "lspci | grep Ethernet | grep \"Virtual Function\" | cut -d ' ' -f1";
+  my @vfs = split("\n", $vfs);
+
+}
+
+
+#get vf device id in libvirt by BDF
+sub get_device_id(){
+  my $device_bdf = shift;
+
+  #change to device id in libvirt
+  $device_bdf =~ s/[:\.]/_/g;
+  #once get nothing while device is listed after a moment
+  my $device_id = script_output "virsh nodedev-list | grep $device_bdf";  
+  return 1 if ! $device_id;
+  $device_id;
+
+}
+
+
+#plugin a device to vm via libvirt
+sub plugin_device_kvm() {
+  my ($vm, $host_device_id) = @_;
+
+  #get neccessary device config from host
+  my $host_device_xml = script_output "virsh nodedev-dumpxml $host_device_id";
+  $host_device_xml =~ /\<domain\>(\w+)\<\/domain\>.*\<bus\>(\w+)\<\/bus\>.*\<slot\>(\w+)\<\/slot\>.*\<function\>(\w+)\<\/function\>/s;
+  my ($dev_domain, $dev_bus, $dev_slot, $dev_func) = ($1, $2, $3, $4);
+  
+  #'virsh attach-interface' can plug the network device to guest with the BDF in commandline without a seperate device xml file
+  #but the SUSE document use 'virsh attach-device', so the test follows it
+  #create the device xml to passthrough to vm
+  assert_script_run("echo \"<interface type='hostdev'>\n  <source>\n    <address type='pci' domain='$dev_domain' bus='$dev_bus' slot='$dev_slot' function='$dev_func'/>\n  </source>\n</interface>\" > $host_device_id.xml", 60);
+
+  #attach device to vm
+  assert_script_run "virsh -d 1 attach-device $vm $host_device_id.xml --persistent";
+  save_screenshot;
+
+  #check the vm xml
+  script_run "virsh dumpxml $vm | grep -A8 interface";
```

Comment:
If there will be confirmed the  explicit return.
imho, script_run() do not work this. Suggestion use  assert_script_run() here.

Reply:
this is not the last statement in the sub function and will not be returned. It is just for test viewers to see the output, more like a debug info, but may be improved to become a check point... I am thinking it over...